### ISSUE TITLE
HTML resources HTTP Caching is broken on webcompat.com

### ISSUE THREAD

**Issue by karlcow** [webcompat/webcompat.com#609]
We need to have a better story at caching HTML resources on webcompat.com.
See #590 and comments on #608

**Comment by miketaylr** [comment]
Yeah, especially since all the dynamic stuff is happening on the client side there's no good reason to not heavily cache the HTML.

**Comment by karlcow** [comment]
Working on this.

**Comment by zoepage** [comment]
@karlcow are you adding AppCache or what is your plan?
(Didn't see anything related to that question in the other linked issues)

**Comment by karlcow** [comment]
@zoepage the issue is about HTTP caching for HTML resources.

Currently if we request an issue
```
→ http --print h GET https://webcompat.com/issues/100
```

The HTTP response has no HTTP caching headers

```http
HTTP/1.1 200 OK
Connection: keep-alive
Content-Encoding: gzip
Content-Type: text/html; charset=utf-8
Date: Thu, 19 Jan 2017 12:48:55 GMT
Server: nginx/1.1.19
Strict-Transport-Security: max-age=31536000
Transfer-Encoding: chunked
Vary: Accept-Encoding
```

I'm adding `Etag` and `Cache-Control`. See the commits already done in that issue, or the tests.
https://github.com/karlcow/webcompat.com/commit/f6a65b7098eb307ef2190296d55f92259e89a1e2

Does it help to understand?

**Comment by karlcow** [comment]
And in case, someone else reading this issue wants to know how HTTP caching is working, there is a very good resource: [Caching Tutorial for Web Authors and Webmasters](http://www.web-caching.com/mnot_tutorial/)

**Comment by karlcow** [comment]
let me restart a branch for this. So I do it against the latest master.

**Comment by miketaylr** [comment]
I'm gonna back this out in https://github.com/webcompat/webcompat.com/issues/1386, because it caused #1384. When we re-land this code, we should verify that #1384 is fixed.

(see also #1385 for writing tests to make sure we don't break logging in/out again)

**Comment by miketaylr** [comment]
Let's also make sure we don't break the test in https://github.com/webcompat/webcompat.com/issues/1371 when we fix it too, I think this changeset caused the permafail (which makes sense, it was a test on an issue page that involved logging in).

**Comment by miketaylr** [comment]
I wonder if the reason it broke is because we serve different HTML if the user is authed:

https://github.com/webcompat/webcompat.com/blob/master/webcompat/templates/shared/shared-nav-links.html#L7-L12

Not sure how to fix that. 🤔

**Comment by karlcow** [comment]
Interesting issue. Let me check it. 😄

**Comment by karlcow** [comment]
On a clean profile. No previous cookies.

```
→ git checkout 609/2
Switched to branch '609/2'
→ nosetests -v
…
Ran 42 tests in 6.603s

OK
→ python run.py
…
```

I go to the homepage on localhost. Click on login. Then boom

```
127.0.0.1 - - [06/Mar/2017 10:57:49] "GET /login HTTP/1.1" 302 -
127.0.0.1 - - [06/Mar/2017 10:57:57] "GET /callback?code=****** HTTP/1.1" 500 -
Traceback (most recent call last):
  File "/Users/karl/.virtualenvs/webcompatcom/lib/python2.7/site-packages/flask/app.py", line 1994, in __call__
    return self.wsgi_app(environ, start_response)
  File "/Users/karl/.virtualenvs/webcompatcom/lib/python2.7/site-packages/flask/app.py", line 1985, in wsgi_app
    response = self.handle_exception(e)
  File "/Users/karl/.virtualenvs/webcompatcom/lib/python2.7/site-packages/flask/app.py", line 1540, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/Users/karl/.virtualenvs/webcompatcom/lib/python2.7/site-packages/flask/app.py", line 1982, in wsgi_app
    response = self.full_dispatch_request()
  File "/Users/karl/.virtualenvs/webcompatcom/lib/python2.7/site-packages/flask/app.py", line 1614, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/Users/karl/.virtualenvs/webcompatcom/lib/python2.7/site-packages/flask/app.py", line 1517, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/Users/karl/.virtualenvs/webcompatcom/lib/python2.7/site-packages/flask/app.py", line 1612, in full_dispatch_request
    rv = self.dispatch_request()
  File "/Users/karl/.virtualenvs/webcompatcom/lib/python2.7/site-packages/flask/app.py", line 1598, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/Users/karl/.virtualenvs/webcompatcom/lib/python2.7/site-packages/flask_github.py", line 183, in decorated
    return f(*((data,) + args), **kwargs)
  File "/Users/karl/code/webcompat.com/webcompat/views.py", line 102, in authorized
    session_db.commit()
  File "/Users/karl/.virtualenvs/webcompatcom/lib/python2.7/site-packages/sqlalchemy/orm/scoping.py", line 149, in do
    return getattr(self.registry(), name)(*args, **kwargs)
  File "/Users/karl/.virtualenvs/webcompatcom/lib/python2.7/site-packages/sqlalchemy/orm/session.py", line 765, in commit
    self.transaction.commit()
  File "/Users/karl/.virtualenvs/webcompatcom/lib/python2.7/site-packages/sqlalchemy/orm/session.py", line 370, in commit
    self._prepare_impl()
  File "/Users/karl/.virtualenvs/webcompatcom/lib/python2.7/site-packages/sqlalchemy/orm/session.py", line 350, in _prepare_impl
    self.session.flush()
  File "/Users/karl/.virtualenvs/webcompatcom/lib/python2.7/site-packages/sqlalchemy/orm/session.py", line 1879, in flush
    self._flush(objects)
  File "/Users/karl/.virtualenvs/webcompatcom/lib/python2.7/site-packages/sqlalchemy/orm/session.py", line 1997, in _flush
    transaction.rollback(_capture_exception=True)
  File "/Users/karl/.virtualenvs/webcompatcom/lib/python2.7/site-packages/sqlalchemy/util/langhelpers.py", line 57, in __exit__
    compat.reraise(exc_type, exc_value, exc_tb)
  File "/Users/karl/.virtualenvs/webcompatcom/lib/python2.7/site-packages/sqlalchemy/orm/session.py", line 1961, in _flush
    flush_context.execute()
  File "/Users/karl/.virtualenvs/webcompatcom/lib/python2.7/site-packages/sqlalchemy/orm/unitofwork.py", line 370, in execute
    rec.execute(self)
  File "/Users/karl/.virtualenvs/webcompatcom/lib/python2.7/site-packages/sqlalchemy/orm/unitofwork.py", line 523, in execute
    uow
  File "/Users/karl/.virtualenvs/webcompatcom/lib/python2.7/site-packages/sqlalchemy/orm/persistence.py", line 64, in save_obj
    mapper, table, insert)
  File "/Users/karl/.virtualenvs/webcompatcom/lib/python2.7/site-packages/sqlalchemy/orm/persistence.py", line 562, in _emit_insert_statements
    execute(statement, multiparams)
  File "/Users/karl/.virtualenvs/webcompatcom/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 717, in execute
    return meth(self, multiparams, params)
  File "/Users/karl/.virtualenvs/webcompatcom/lib/python2.7/site-packages/sqlalchemy/sql/elements.py", line 317, in _execute_on_connection
    return connection._execute_clauseelement(self, multiparams, params)
  File "/Users/karl/.virtualenvs/webcompatcom/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 814, in _execute_clauseelement
    compiled_sql, distilled_params
  File "/Users/karl/.virtualenvs/webcompatcom/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 927, in _execute_context
    context)
  File "/Users/karl/.virtualenvs/webcompatcom/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 1076, in _handle_dbapi_exception
    exc_info
  File "/Users/karl/.virtualenvs/webcompatcom/lib/python2.7/site-packages/sqlalchemy/util/compat.py", line 185, in raise_from_cause
    reraise(type(exception), exception, tb=exc_tb)
  File "/Users/karl/.virtualenvs/webcompatcom/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 920, in _execute_context
    context)
  File "/Users/karl/.virtualenvs/webcompatcom/lib/python2.7/site-packages/sqlalchemy/engine/default.py", line 425, in do_execute
    cursor.execute(statement, parameters)
IntegrityError: (IntegrityError) datatype mismatch u'INSERT INTO users (user_id, access_token) VALUES (?, ?)' ('******', u'*******')
127.0.0.1 - - [06/Mar/2017 10:57:57] "GET /callback?__debugger__=yes&cmd=resource&f=style.css HTTP/1.1" 200 -
127.0.0.1 - - [06/Mar/2017 10:57:57] "GET /callback?__debugger__=yes&cmd=resource&f=jquery.js HTTP/1.1" 200 -
127.0.0.1 - - [06/Mar/2017 10:57:57] "GET /callback?__debugger__=yes&cmd=resource&f=debugger.js HTTP/1.1" 200 -
127.0.0.1 - - [06/Mar/2017 10:57:57] "GET /callback?__debugger__=yes&cmd=resource&f=console.png HTTP/1.1" 200 -
127.0.0.1 - - [06/Mar/2017 10:57:57] "GET /callback?__debugger__=yes&cmd=resource&f=ubuntu.ttf HTTP/1.1" 200 -
127.0.0.1 - - [06/Mar/2017 10:57:57] "GET /callback?__debugger__=yes&cmd=resource&f=console.png HTTP/1.1" 200 -
127.0.0.1 - - [06/Mar/2017 10:58:02] "GET /callback?__debugger__=yes&cmd=resource&f=source.png HTTP/1.1" 200 -
```

hmm interesting. Let's stop the server. Quit the browser. Restart with a fresh profile.

```
rm -rf session.db
```

* This time it is working from the home page. 
* Logout. 
* Login. Working. 
* Switch to an issue view. 
* And try to logout. It doesn't work.

Let's see the network panel.

The logout doesn't seem to have issues. It is recorded. with a the appropriate location.

```http
HTTP/1.0 302 FOUND
Content-Type: text/html; charset=utf-8
Content-Length: 271
Location: http://localhost:5000/issues/522
Set-Cookie: session=********; HttpOnly; Path=/
Server: Werkzeug/0.10.4 Python/2.7.10
Date: Mon, 06 Mar 2017 02:05:21 GMT
```

The request headers didn't have anything related to caching

```http
GET /logout HTTP/1.1
Host: localhost:5000
User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10.12; rv:54.0) Gecko/20100101 Firefox/54.0
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8
Accept-Language: en-US,en;q=0.5
Accept-Encoding: gzip, deflate
Referer: http://localhost:5000/issues/522
Cookie: session=*****
Connection: keep-alive
Upgrade-Insecure-Requests: 1
````

What is happening? Why is it working on the home page and not on the issue page. 

Ah gotcha! It DOES login or logout.
What is not working is the refresh on the status bar. So it seems to be the JS part.

**Comment by karlcow** [comment]
So indeed. This is a templating issue. 
And to make we don't loose the lines of code.
https://github.com/webcompat/webcompat.com/blob/5d7604bdeee7a25137e35fef47ab767f4574f7f6/webcompat/templates/shared/shared-nav-links.html#L7-L12

```python
{% if not session.user_id %}
  <a class="wc-Navbar-link js-login-link" href="{{ url_for('login') }}">
    <span class="wc-Navbar-link-icon wc-Icon wc-Icon--arrow-circle-right" aria-hidden="true"></span>
    <span class="wc-Navbar-link-label">Log in</span>
  </a>
{% else %}
  <div class="r-ResetButton wc-Navbar-link js-login-link wc-DropdownHeader">
    <button class="r-ResetButton" aria-pressed="false">
      <img class="wc-Navbar-avatar" src="{{ session.avatar_url }}s=50" alt="User Avatar">
      <span class="wc-Navbar-link-icon wc-Icon wc-Icon--chevron-down" aria-hidden="true"></span>
    </button>
    <div class="wc-DropdownHeader-content">
      <a class="wc-Navbar-linkDropdown" href="{{ url_for('me_redirect') }}">My Activity</a>
      <a class="wc-Navbar-linkDropdown" href="{{ url_for('logout') }}">Logout</a>
    </div>
  </div>
{% endif %}
```

**Comment by karlcow** [comment]
I need to test a couple of things. Because the content should be changing, hence the etag.

**Comment by karlcow** [comment]
Let get's to http://localhost:5000/issues/200 with a fresh profile
No login. No cookies.

```http
GET /issues/515 HTTP/1.1
Host: localhost:5000
User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10.12; rv:54.0) Gecko/20100101 Firefox/54.0
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8
Accept-Language: en-US,en;q=0.5
Accept-Encoding: gzip, deflate
Referer: http://localhost:5000/
Cookie: ***
Connection: keep-alive
Upgrade-Insecure-Requests: 1
```

and Response is:

```http
HTTP/1.0 200 OK
Content-Type: text/html; charset=utf-8
Content-Length: 9366
Cache-Control: private, max-age=86400
Etag: "87f1ed4981ad4a15ae5a2fe84f250ac2"
Date: Mon, 06 Mar 2017 04:41:43 GMT
Server: Werkzeug/0.10.4 Python/2.7.10
```

This is normal.

Let's do a reload

```http
GET /issues/515 HTTP/1.1
Host: localhost:5000
User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10.12; rv:54.0) Gecko/20100101 Firefox/54.0
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8
Accept-Language: en-US,en;q=0.5
Accept-Encoding: gzip, deflate
Referer: http://localhost:5000/issues?page=1&per_page=50&state=open&stage=needstriage&sort=created&direction=desc
Cookie: session=****
Connection: keep-alive
Upgrade-Insecure-Requests: 1
If-None-Match: "87f1ed4981ad4a15ae5a2fe84f250ac2"
Cache-Control: max-age=0
```

Then the response is:

```http
HTTP/1.0 304 NOT MODIFIED
Cache-Control: private, max-age=86400
Etag: "87f1ed4981ad4a15ae5a2fe84f250ac2"
Date: Mon, 06 Mar 2017 04:45:07 GMT
Connection: close
Server: Werkzeug/0.10.4 Python/2.7.10
```

Working!

Ok let's click the login button. We know that the login is working, but the page doesn't show the new page with the different template.

The request headers according to developer tools (I'm dubious says)

```http
GET /issues/515 HTTP/1.1
Host: localhost:5000
User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10.12; rv:54.0) Gecko/20100101 Firefox/54.0
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8
Accept-Language: en-US,en;q=0.5
Accept-Encoding: gzip, deflate
Referer: https://github.com/
Cookie: session=*****
Connection: keep-alive
Upgrade-Insecure-Requests: 1
```

and the response headers

```http
Content-Type: text/html; charset=utf-8
Content-Length: 9366
Cache-Control: private, max-age=86400
Etag: "87f1ed4981ad4a15ae5a2fe84f250ac2"
Server: Werkzeug/0.10.4 Python/2.7.10
Date: Mon, 06 Mar 2017 04:45:07 GMT
```

but developer tools says also that it comes from the cache and indeed the cookie session is not in there for the response and on the server side log

```
127.0.0.1 - - [06/Mar/2017 13:47:27] "GET /login HTTP/1.1" 302 -
127.0.0.1 - - [06/Mar/2017 13:47:40] "GET /callback?code=***** HTTP/1.1" 302 -
127.0.0.1 - - [06/Mar/2017 13:47:42] "GET /api/issues/labels?per_page=100 HTTP/1.1" 200 -
127.0.0.1 - - [06/Mar/2017 13:47:43] "GET /api/issues/515 HTTP/1.1" 200 -
127.0.0.1 - - [06/Mar/2017 13:47:43] "GET /api/issues/515/comments?page=1 HTTP/1.1" 200 -
````

There is no request to… the issue URI. So the request is going through.

if I do another force reload my login state is shown. 

```http
GET /issues/515 HTTP/1.1
Host: localhost:5000
User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10.12; rv:54.0) Gecko/20100101 Firefox/54.0
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8
Accept-Language: en-US,en;q=0.5
Accept-Encoding: gzip, deflate
Referer: https://github.com/
Cookie: session=****
Connection: keep-alive
Upgrade-Insecure-Requests: 1
If-None-Match: "87f1ed4981ad4a15ae5a2fe84f250ac2"
Cache-Control: max-age=0
```

There's the correct `If-None-Match: "87f1ed4981ad4a15ae5a2fe84f250ac2"`

and the response is

```http
HTTP/1.0 200 OK
Content-Type: text/html; charset=utf-8
Content-Length: 11944
Cache-Control: private, max-age=86400
Etag: "a13d8f68a2abc2ae1c0d7c588dd2db82"
Date: Mon, 06 Mar 2017 04:54:18 GMT
Set-Cookie: *****
Server: Werkzeug/0.10.4 Python/2.7.10
```

with a different etag as it should be because it's based on the HTML content and the template changed.

So the caching is working properly, the login dance in between github/webcompat works for the cookie but doesn't generate the correct request with a if-none-match. 

Basically No conditional request is done because the resource is aged less than 1 day so it is using the cache.

**Comment by karlcow** [comment]
I have two ideas to fix it. Let's try.

**Comment by karlcow** [comment]
So basically the right thing to use here would be `no-cache` to force revalidation but it seems that Firefox and IE at least treat `no-cache` as `no-store`. So that's not good.

The way to do something which is interoperable is to do.

`must-revalidate, max-age=0` Basically we ask the browser to force the 304 Not Modified response from the server. So it's still a hit on the server (like the 200 before), but at least it's a hit without a body when the response has not been modified, so it's less traffic than the 200.

**Comment by karlcow** [comment]
I will create a new branch with the appropriate information.

**Comment by karlcow** [comment]
```
127.0.0.1 - - [09/Mar/2017 15:14:35] "GET / HTTP/1.1" 200 -

127.0.0.1 - - [09/Mar/2017 15:14:42] "GET /issues/515 HTTP/1.1" 200 -
127.0.0.1 - - [09/Mar/2017 15:14:43] "GET /api/issues/515 HTTP/1.1" 200 -

127.0.0.1 - - [09/Mar/2017 15:15:08] "GET /issues/515 HTTP/1.1" 304 -
127.0.0.1 - - [09/Mar/2017 15:15:14] "GET /api/issues/515 HTTP/1.1" 304 -

# LOGIN
127.0.0.1 - - [09/Mar/2017 15:15:48] "GET /login HTTP/1.1" 302 -
127.0.0.1 - - [09/Mar/2017 15:15:50] "GET /callback?code=**** HTTP/1.1" 302 -
127.0.0.1 - - [09/Mar/2017 15:15:51] "GET /issues/515 HTTP/1.1" 200 -

# RELOAD
127.0.0.1 - - [09/Mar/2017 15:16:31] "GET /issues/515 HTTP/1.1" 304 -
127.0.0.1 - - [09/Mar/2017 15:16:38] "GET /api/issues/515 HTTP/1.1" 200 -

# LOGOUT
127.0.0.1 - - [09/Mar/2017 15:17:14] "GET /logout HTTP/1.1" 302 -
127.0.0.1 - - [09/Mar/2017 15:17:14] "GET /issues/515 HTTP/1.1" 200 -

# LOGIN
127.0.0.1 - - [09/Mar/2017 15:17:46] "GET /login HTTP/1.1" 302 -
127.0.0.1 - - [09/Mar/2017 15:17:53] "GET /callback?code=***** HTTP/1.1" 302 -
127.0.0.1 - - [09/Mar/2017 15:17:54] "GET /issues/515 HTTP/1.1" 200 -

#RELOAD
127.0.0.1 - - [09/Mar/2017 15:18:01] "GET /issues/515 HTTP/1.1" 304 -
127.0.0.1 - - [09/Mar/2017 15:18:02] "GET /api/issues/515 HTTP/1.1" 200 -


#LOGOUT
127.0.0.1 - - [09/Mar/2017 15:18:38] "GET /logout HTTP/1.1" 302 -
127.0.0.1 - - [09/Mar/2017 15:18:38] "GET /issues/515 HTTP/1.1" 200 -
```

**Comment by miketaylr** [comment]
This is fixed.
